k2/csrc/determinize.h

```diff
@@ -0,0 +1,686 @@
+  in the determinized output.  Caution: it uses a randomized algorithm that
+  could in principle produce collisions that would generate wrong output.
+  We don't think this will ever happen though (128-bit representations).
+ */
+template <class TracebackState>
+class DetStateMap {
+ public:
+  /*
+    Looks up the output state-id corresponding to a specific DetState structure,
+    creating a new output-state if necessary.  This does not store any pointers
+    to the DetState or its contents, so you can delete the DetState without
+    affecting this object's ability to map an equivalent DetState to the same
+    state-id.
+
+       @param [in,out] a  The DetState whose state-id we are looking up.
+                         The integer id of the output-FSA state will be written
+                         to its `output_state` field, which at entry is assumed
+                         to be unset.
+        @return  Returns true if this was a NEWLY CREATED state,
+              false otherwise.
+   */
+  bool GetOutputState(DetState<TracebackState> *a) {
+    std::pair<uint64_t, uint64_t> compact;
+    DetStateToCompact(a, &compact);
+    auto p = map_.insert({compact, cur_output_state_});
+    bool inserted = p.second;
+    if (inserted) {
+      a->state_id = cur_output_state_++;
+      return true;
+    } else {
+      a->state_id = p.first->second;
+      return false;
+    }
+  }
+
+  int32_t size() const { return cur_output_state_; }
+
+ private:
+  // simple hashing function that just takes the first element of the pair.
+  struct PairHasher {
+    size_t operator()(const std::pair<uint64_t, uint64_t> &p) const {
+      return static_cast<size_t>(p.first);
+    }
+  };
+
+  int32_t cur_output_state_{0};
+  /* maps from 128-bit key (stored as a pair of uint64_t's) to the int32_t
+   * state-id. */
+  std::unordered_map<std::pair<uint64_t, uint64_t>, uint32_t, PairHasher> map_;
+
+  /* Turns DetState into a compact form of 128 bits.  Technically there
+     could be collisions, which would be fatal for the algorithm, but this
+     is one of those lifetime-of-the-universe type of things (kind of like
+     the theoretical potential for git hash collision) that we ignore.
+
+     The normalized form
+
+  */
+  void DetStateToCompact(const DetState<TracebackState> &d, const Fsa &fsa,
+                         std::pair<uint64_t, uint64_t> *vec) {
+    assert(d.normalized);
+
+    uint64_t a = d.base_state + 17489 * d.seq_len,
+             b = d.base_state * 103979 + d.seq_len;
+
+    // We choose an arbitrary DetStateElement (the first one in the list) to
+    // read the symbol sequence from; the symbol sequence will be the same no
+    // matter which element we choose to trace back.
+    auto elem = d.elements.begin()->second;
+    int32_t seq_len = d.seq_len;
+    const auto &arcs = fsa.arcs;
+    for (int32_t i = 0; i < seq_len; ++i) {
+      int32_t symbol = arcs[elem->arc_id].label;
```

Comment:
as `MaxTracebackState` and `LogSumTracebackState` has differnt structure, we may need to define differnt functions for them, For `LogSum`, how do compute the hash pair as there are a vector of prev_elements `vector<LogSumTracebackLink> prev_elements`, iterover all elements in it?

Reply:
No you can just pick an arbitrary one, the symbol-sequence will be the same no matter which you pick.

Reply:
OK, I just picked the first one.